### DISCUSSION TITLE
Improve enums and make them more modern with different syntax, enum infer, more enum types

### DISCUSSION THREAD

**Discussion by Dynamic-Pistol** [godotengine/godot-proposals#10833] [Scripting]
### Describe the project you are working on

Nothing specific, but multiple projects in general could use this and it makes it more consistent with godot's syntax,making the language more modern and specifically the 3rd proposal would be a new feature,these could also be separated into their own proposals if they get enough positive feedback here

### Describe the problem or limitation you are having in your project

There are 3 major issues with enums in godot:

1. Gdscript enums are declared differently than classes and functions
2. No enum inference for gdscript
3. Terrible flags and bitfield usage

additional issues that can be made into their own proposal due to being too big or are just not important enough to list
- Making enums have functions
- Changing recommended enum syntax
- Making enums not behave like constant dictionaries

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

1. Make gdscript enums more consistent with other declarations
2. Add inferring gdscript enums based on context
3. Add `@flags` and `@bitset` annotations,with flags making them have enumflags behavior,and bitsets allowing for better bit usage

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

# 1st: 
```gdscript
enum Enum:
    Lit1,
    Lit2,
    ...
```

# 2nd: 
```gdscript
var e: Enum = .Lit1
```

# 3rd
```gdscript
@flags
enum DamageType{
    Fire,
    Venom,
    Magic,
}

var flag: DamageType = .Fire | .Venom

@bitset
enum StatusEffect{
    Burning,
    Healing,
    Cursed,
}

var bit_set : StatusEffect = .Burning
bit_set += .Healing
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

1. technically, yes, but it's inconsistent with godot's syntax
2. writing the entire enum name + optionally the containing class name,which wouldn't be very good and may be unreadable in certain cases
3. Godot flags are already there,but they require bitwise operations which maybe confusing at times for beginners, and for bitsets,they don't exist

### Is there a reason why this should be core and not an add-on in the asset library?

Enums are a core part of the gdscript module of godot.

**Comment by AThousandShips** [comment]
These are absolutely three distinct proposals that would make much more sense to discuss separately

**Comment by Dynamic-Pistol** [comment]
@AThousandShips i think they are small enough and related enough to discuss first here,then i can make a new proposal when enough feedback for one is made.

**Comment by KoBeWi** [comment]
The first one was proposed in #4278

**Comment by jmarceno** [comment]
> @AThousandShips i think they are small enough and related enough to discuss first here,then i can make a new proposal when enough feedback for one is made.

So, shouldn't it be a discussion? 
As I understand this is what they are for.

**Comment by Dynamic-Pistol** [comment]
> So, shouldn't it be a discussion? As I understand this is what they are for.

tbh i forgot those existed,but also i asked on the chat instance and it was said if they were small enough,it would be okay to post them all here,but if it's possible to convert this to a discussion,i wouldn't mind

**Comment by Meorge** [comment]
Enums closer to this (especially the dot notation shorthand, your second recommendation) are something I loved in Swift and would love to see in GDScript.  

> writing the entire enum name + optionally the containing class name,which wouldn't be very good and may be unreadable in certain cases

I think this may already be somewhat how enums are suggested to be handled in GDScript. For example, I work a lot with tweens, and I know that [transition type enum values](https://docs.godotengine.org/en/stable/classes/class_tween.html#enum-tween-transitiontype) all start with a common prefix: `TRANS_LINEAR`, `TRANS_SINE`, `TRANS_QUAD`, etc.
It'd be a pretty major breaking change, so maybe something to consider for Godot 5 haha, but I would personally love something closer to Swift, like
```gdscript
tween.set_trans(.sine)
```
over
```gdscript
tween.set_trans(Tween.TRANS_SINE)
# or
tween.set_trans(Tween.TransitionType.TRANS_SINE)
```
